nixos-rebuild: Add list-generations

miallo opened the pull request:
## Description of changes

Add new option `nixos-rebuild list-generations`. It will show an output like

```
$ nixos-rebuild list-generations
Generation      Build-date               NixOS version           Kernel  Configuration Revision                    Specialisations
52   (current)  Fri 2023-08-18 08:17:27  23.11.20230817.0f46300  6.4.10  448160aeccf6a7184bd8a84290d527819f1c552c  *
51              Mon 2023-08-07 17:56:41  23.11.20230807.31b1eed  6.4.8   99ef480007ca51e3d440aa4fa6558178d63f9c42  *
```
or in a more machine readable format with `--json`:
```json
[
  {
    "generation": 52,
    "date": "2023-08-18T06:17:27Z",
    "nixosVersion": "23.11.20230817.0f46300",
    "kernelVersion": "6.4.10",
    "configurationRevision": "448160aeccf6a7184bd8a84290d527819f1c552c",
    "specialisations": [
      "*"
    ],
    "current": true
  },
  {
    "generation": 51,
    "date": "2023-08-07T15:56:41Z",
    "nixosVersion": "23.11.20230807.31b1eed",
    "kernelVersion": "6.4.8",
    "configurationRevision": "99ef480007ca51e3d440aa4fa6558178d63f9c42",
    "specialisations": [
      "*"
    ],
    "current": false
  }
]
```


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

This is actually the first commit of https://github.com/NixOS/nixpkgs/pull/105910 because it does not really make sense to couple the two changes into one single merge-request and so this is hopefully easier to review. I am sorry for doing it like this in the first place :see_no_evil:

@Mic92 @SuperSandro2000 and @xaverdh already had a first look at the combined merge request a while ago, hopefully it is easier now to review this one change that is also not as critical as the second part.

What is missing that anyone would like to have in the overview?

miallo committed:
nixos-rebuild: Add list-generations

Add new command `nixos-rebuild list-generations`. It will show an output
like

```
$ nixos-rebuild list-generations
Generation      Build-date               NixOS version           Kernel  Configuration Revision                    Specialisations
52   (current)  Fri 2023-08-18 08:17:27  23.11.20230817.0f46300  6.4.10  448160aeccf6a7184bd8a84290d527819f1c552c  *
51              Mon 2023-08-07 17:56:41  23.11.20230807.31b1eed  6.4.8   99ef480007ca51e3d440aa4fa6558178d63f9c42  *
```

This also mentions the change in the upcoming release notes

SuperSandro2000 commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -507,6 +511,87 @@ if [ "$action" = dry-build ]; then
+        # jq automatically quotes the output => don't try to quote it in output!
+        build_date="$(stat "$generation_dir" --format=%W | jq 'todate')"
+
+        pushd "$generation_dir/specialisation/" > /dev/null || :;
+        specialisation_list=(*)
+        popd > /dev/null || :;
```
```suggestion
        pushd "$generation_dir/specialisation/" > /dev/null || :
        specialisation_list=(*)
        popd > /dev/null || :
```

miallo commented:
@Mic92 and @xaverdh - you already reviewed an outdated version of this as part of #105910 a long time ago. Do you want to have a look at it in this cleaner form (and with only a single feature per PR)? Not sure if Eelco also should be involved, or if that was more focused on the second feature (switching to a specific generation) that aroused his interest.

Obviously all commits should be squashed before merging. I just left them like this for now for better reviewability/transparency.

Sidenote: with many generations this is relatively slow right now (60 generations = ~10s), but with the next `jq`-version from #246515 there is roughly a 3x performance increase, so it becomes at least sort of responsive... I wonder how I can reduce the number of `jq`-calls or which other performance improvements I can make further 🤔

doronbehar commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.8:
```diff
@@ -10,7 +10,7 @@
 .Nm
 .Bro
-.Cm switch | boot | test | build | dry-build | dry-activate | edit | build-vm | build-vm-with-bootloader
+.Cm switch | boot | test | build | dry-build | dry-activate | edit | build-vm | build-vm-with-bootloader | list-generations Op Fl -json
```
I'm not experienced at all with manpages writing, but for sure this shouldn't be this:

```suggestion
.Cm switch | boot | test | build | dry-build | dry-activate | edit | build-vm | build-vm-with-bootloader | list-generations Op Fl --json
```

?

miallo replied:
I think the second "-" is taken care of by the `Fl`. At least it is shown properly as `--json` when I run `man pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.8`. I have to say though, that I am totally unfamiliar with this man-page syntax...

doronbehar replied:
> At least it is shown properly as `--json` when I run `man pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.8`.

I trust you on that :+1:.

miallo replied:
You don't have to: After having built the whole thing many times from scratch I read somewhere that you can just run `man <path to file.8>`. Makes life a lot easier when dealing with this "interesting" syntax 😂 

Actually now that I think about it: my very first version of this (at least the one that was not the old XML doc) also had the same code you suggested, but that showed `---json` and I fixed it later ^^

doronbehar commented:
Very nice! Much better approach implementation then #242926 IMO.

doronbehar commented:
This seems pretty ready to me overall, good to merge after a squash. It had been given enough :+1:, and will have more then a single approval soon.

doronbehar commented:
OK, please squash the commits, I'll run a `nixpkgs-review`, and I'll merge this.

miallo commented:
@doronbehar Should I add an entry in [the release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md)? Which section - Highlights? Should it be the commit message, i.e.:
Add  new command `nixos-rebuild list-generations`. It will show an output like
```
$ nixos-rebuild list-generations
Generation      Build-date               NixOS version           Kernel  Configuration Revision                    Specialisations
52   (current)  Fri 2023-08-18 08:17:27  23.11.20230817.0f46300  6.4.10  448160aeccf6a7184bd8a84290d527819f1c552c  *
51              Mon 2023-08-07 17:56:41  23.11.20230807.31b1eed  6.4.8   99ef480007ca51e3d440aa4fa6558178d63f9c42  *
```

doronbehar commented:
> @doronbehar Should I add an entry in [the release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md?rgh-link-date=2023-09-10T12%3A57%3A31Z)?

Yes why not.

> Which section - Highlights?

Definitely good idea, not in the commit message only. I'd do something like this:

In `rl-2311.section.md`:

```markdown
The `nixos-rebuild` command has been given a `list-generations` subcommand. See `man nixos-rebuild` for more details.
```

Squash it all to a single commit, and mention in the commit that the release notes were added.

miallo commented:
> The `nixos-rebuild` command has been given a `list-generations` subcommand. See `man nixos-rebuild` for more details.

A lot better than my suggestion!

> Squash it all to a single commit, and mention in the commit that the release notes were added.

I did. The only other thing I fixed was that I just saw an inconsistency in the man page where I had exactly the `Op --json` instead of `Op Fl -json`. I changed this, so it is in line with all the other places of this man page
